### PR TITLE
RFC: Reduce the number of db-config-* flags from 60 to 20.

### PR DESCRIPTION
**Pull request by jvaidya**
Currently, we support 5 different connection types: "App", "Dba", "Allprivs", "Filtered" and "Repl".
For each of these, we create 12 flags, one for each field of mysql.ConnParams struct.

The (perhaps naive) assumption behind this CR is that since we are creating connections to the same mysqld instance, the only parameters that need to be different across these connection types are username and password.

I would like to get comments on this CR.

### CODE REVIEW

**Comment by jvaidya** [comment]
I signed it!

File: py/vttest/vt_processes.py [hunk 1]
```diff
@@ -160,14 +159,13 @@ def __init__(self, directory, topology, mysql_db, schema_dir, charset,
...
-          ['-db-config-app-unixsocket', mysql_db.unix_socket(),
-           '-db-config-dba-unixsocket', mysql_db.unix_socket()])
+          ['-db-config-unixsocket', mysql_db.unix_socket()])
     else:
       self.extraparams.extend(
-          ['-db-config-app-host', mysql_db.hostname(),
-           '-db-config-app-port', str(mysql_db.port()),
-           '-db-config-dba-host', mysql_db.hostname(),
-           '-db-config-dba-port', str(mysql_db.port())])
+          ['-db-config-host', mysql_db.hostname(),
```
Comment (jvaidya) [inline py/vttest/vt_processes.py]: Thanks for catching. Fixed.

File: test/tablet.py [hunk 2]
```diff
@@ -687,15 +688,13 @@ def _add_dbconfigs(self, cfg, args, repl_extra_flags=None):
...
-        config['app']['dbname'] = self.dbname
-      if 'repl' in config:
-        config['repl']['dbname'] = self.dbname
+      config['global']['dbname'] = self.dbname
     if 'repl' in config:
       config['repl'].update(repl_extra_flags)
     for key1 in config:
       for key2 in config[key1]:
-        args.extend(['-db-config-' + key1 + '-' + key2, config[key1][key2]])
+        fkey = '' if key1 == 'global' else '%s-' % key1
```
Comment (jvaidya) [inline test/tablet.py]: Much better. Changed.

File: go/vt/dbconfigs/dbconfigs.go [hunk 3]
```diff
@@ -156,32 +164,51 @@ func (dbcfgs *DBConfigs) IsZero() bool {
...
+	toConnParams.SslKey = fromConnParams.SslKey
+}
+
 // Init will initialize app, allprivs, dba, filtered and repl configs.
 func Init(socketFile string, flags DBConfigFlag) (*DBConfigs, error) {
 	if flags == EmptyConfig {
 		panic("No DB config is provided.")
 	}
 	if AppConfig&flags != 0 {
+		copyConnectionFields(&globalConnParams, &dbConfigs.App)
```
Comment (jvaidya) [inline go/vt/dbconfigs/dbconfigs.go]: Good idea. Moved.

**Comment by jvaidya** [comment]
> Can you please check for other occurrences as well?

I have fixed all occurances except in '/docs'.

**Comment by jvaidya** [comment]
The tarvis tests are failing. The cause seems to be that vttablet is being
passed "-db-config-repl-{ssl-key,flags,ssl-cert,ssl-ca}".

@michael-berlin, @alianjobart: I have 2 questions:

1. I can not find where in source code these flags are coming from. Do
these come from premade docker containers or some such?

2. Is this actually a legitimate usecase that my CR does not cover? That
is, "repl" might need connection parameters that are different from the
other types?


https://travis-ci.org/youtube/vitess/jobs/246628750#L555


2017-06-24 20:23:41,993 utils:306 DEBUG run:
['/home/travis/gopath/bin/vttablet', '-tablet_hostname', 'localhost',
'-tablet-path', 'test_nj-0000062345', '-topo_implementation', 'zk2',
'-topo_global_server_address', 'localhost:16408', '-topo_global_root',
'/global', '-binlog_player_protocol', 'grpc', '-tablet_manager_protocol',
'grpc', '-tablet_protocol', 'grpc',
'-binlog_player_healthcheck_topology_refresh', '1s',
'-binlog_player_healthcheck_retry_delay', '1s',
'-binlog_player_retry_delay', '1s', '-pid_file',
'/home/travis/gopath/vtdataroot/vt_182587236/vt_0000062345/vttablet.pid',
'-health_check_interval', '2s', '-enable_replication_reporter',
'-degraded_threshold', '5s', '-watch_replication_stream',
'-enable_semi_sync', '-restore_from_backup',
'-backup_storage_implementation', 'file', '-file_backup_storage_root',
'/home/travis/gopath/vtdataroot/vt_182587236/tmp/backupstorage', '-port',
'16403', '-log_dir', '/home/travis/gopath/vtdataroot/vt_182587236/tmp',
'-db-config-allprivs-uname', 'vt_allprivs', '-db-config-repl-uname',
'vt_repl', '-db-config-repl-ssl-key',
'/home/travis/gopath/vtdataroot/vt_182587236/tmp/certs/client-key.pem',
'-db-config-repl-flags', '2048', '-db-config-repl-ssl-cert',
'/home/travis/gopath/vtdataroot/vt_182587236/tmp/certs/client-cert.pem',
'-db-config-repl-ssl-ca',
'/home/travis/gopath/vtdataroot/vt_182587236/tmp/certs/ca-cert.pem',
'-db-config-dba-uname', 'vt_dba', '-db-config-filtered-uname',
'vt_filtered', '-db-config-app-uname', 'vt_app', '-db-config-charset',
'utf8', '-db-config-dbname', 'vt_test_keyspace', '-service_map',
'grpc-tabletmanager,grpc-throttler,grpc-queryservice,grpc-updatestream,grpc-vtctl,grpc-vtworker,grpc-vtgateservice',
'-grpc_port', '16405', '-serving_state_grace_period', '1s',
'-enable-autocommit'] extra_env={'EXTRA_MY_CNF': '/home/travis/gopath/src/
github.com/youtube/vitess/config/mycnf/default-fast.cnf:/home/travis/gopath/src/github.com/youtube/vitess/config/mycnf/master_mariadb.cnf'},
stderr=<open file
'/home/travis/gopath/vtdataroot/vt_182587236/tmp/vttablet-62345.stderr',
mode 'w' at 0x2965d20>

**Comment by alainjobart** [comment]
The 'repl' group of parameters are the ones sent to MySQL when we re-parent, to connect to the master. They are not used to connect to the local MySQL, but for the local MySQL to connect to the MySQL master. The host and port come from the topology (the current master), the user/password/SSL config from the repl user. Since this goes over the network, it makes sense to have different parameters there in some cases. So I would keep it as a separate group.

**Comment by jvaidya** [comment]
Thanks @alainjobart. I will fix this in the next couple of days.

**Comment by sougou** [comment]
PR outdated. We'll need to rebuild from scratch.

**Pull request closed without merging**